### PR TITLE
[serapi] Support Coq 8.14

### PR DESCRIPTION
**Pull request by ejgallego**
Nothing significant it seems, other than a bit of a mess with
initialization due to upstream changes. There is much room for
improvement if we get time to refactor. We also have to account for
`coq-core` until we improve plugin location using `findlib`.

### CODE REVIEW

**Comment by ejgallego** [comment]
Working !

**Comment by palmskog** [comment]
Based on discussion in coq-community/docker-coq#29, the current plan is to add a `coq-serapi` package that works with Coq 8.14+rc1 to the `extra-dev` Coq opam repo with the bound `= "8.14+rc1"` on Coq. This enables putting SerAPI in Docker images with Coq 8.14 and testing the various platform packages and other projects that may want to be compatible with 8.14 and might use Alectryon. Once 8.14.0 is out, the SerAPI package is put in the usual opam repo and the `extra-dev` package would be removed.

How does that plan sound to you @ejgallego? And is there anything standing in the way for merging this PR and making a release/tag?

**Comment by ejgallego** [comment]
Note that this PR won't be merged, but we will just change the default branch in the github interface.

**Comment by palmskog** [comment]
@ejgallego I assumed we were working towards both (1) SerAPI becoming part of the Coq Docker images and (2) SerAPI joining the Coq Platform. If your position is that there can't be a SerAPI tag until there can be a release in the main opam repo, then there is no hope for SerAPI ever joining the Coq Platform. As outlined in the release cycle CEP 52, all tagging happens during the X+rc1 phase, and they become the basis for both the basic Docker image (bignums, SerAPI) and the platform Docker image (bignums, SerAPI, Equations, ...).

It seems shaky to me to just pick a SHA in a SerAPI opam package for the basic Docker image. @erikmd what's your position on this?

**Comment by erikmd** [comment]
I guess I understand why Emilio won't merge this PR (because it's just a "heads-up" PR to exemplify the changes from 8.13→8.14)

> @erikmd what's your position on this?

I don't know what is coq-serapi's project tagging policy (@ejgallego might want to elaborate on this a bit if need be)

but technically speaking, it is possible for me to integrate an opam package such as coq-serapi in Docker-Coq as long as:

1. there is a package in coq-"released" or coq-"extra-dev" or the "default" opam-repository;
2. **or** there is a `*.opam` file directly committed in the git repository, allowing one to git-pin a tag or branch or SHA1.

So if you think that 1. is too much hassle, I'm fine with 2., provided:
* the file https://github.com/ejgallego/coq-serapi/blob/v8.14/coq-serapi.opam is updated (it still mentions 8.13, and I believe the [`!pinned`](https://github.com/ejgallego/coq-serapi/blob/d691ddca5931e125ad5deef61d75446972a21a45/coq-serapi.opam#L27) predicate (that wasn't [there](https://github.com/ocaml/opam-repository/blob/master/packages/coq-serapi/coq-serapi.8.13.0%2B0.13.0/opam)) should be removed, shouldn't it?)
* a tag is pushed (maybe in the form of a prelease) in https://github.com/ejgallego/coq-serapi/tags, since:
  * using a branch might not be stable/precise enough for an image involving Coq 8.14+rc1,
  * using a SHA1 is not enough with opam's git pinning, as we also need an opam version string to do something like `opam pin -n -y -k coq-serapi.$(SERAPI_VERSION_STRING) $(SERAPI_GIT_URL_WITH_SHA1)` (see e.g. [this thread](https://stackoverflow.com/a/62503722/9164010)); so if it is needed to provide `SERAPI_VERSION_STRING`, one could just as well use this version string as a tag…
* anyway, I guess a proper version string should be found (I don't know which) that will be almost the same as coq-serapi's version string for the package associated with the final release with Coq 8.14.0, but still different(?)
* I also Cc @Zimmi48 in case he'd like to comment w.r.t. coq-bignums' tagging strategy as well :)

**Comment by Zimmi48** [comment]
The way you manage your branches is up to you, but indeed we do need platform packages to be released (== tags to be set) during the rc period as per CEP 52 that @palmskog referred to.

>For me a release here, is something that is available in the main opam repos.

I personally don't get why coq-serapi is released in the main opam repo contrary to virtually all other Coq packages, but never mind that. If that's your definition of a release, then that's not what we're asking. We just need a tag.

**Comment by ejgallego** [comment]
Umm, maybe I am confused, but I see nothing in https://github.com/coq/ceps/blob/master/text/052-platform-release-cycle.md that requires stuff in the CI to do a tag or a release.

On the contrary, the beta process works over pure pins from CI isn't it the case?

A tag is not a big deal, tho I'd be arbitrary as the others are IIUC.

I usually release SerAPI _after_ the Coq package is in standard opam, so I can properly test it; otherwise it is difficult to know if once Coq v8.14.0 hits opam our package will still work [serapi is quite complex].

> I personally don't get why coq-serapi is released in the main opam repo contrary to virtually all other Coq packages, but never mind that.

What would we gain from that?

**Comment by erikmd** [comment]
> I usually release SerAPI after the Coq package is in standard opam, so I can properly test it; otherwise it is difficult to know if once Coq v8.14.0 hits opam our package will still work

Indeed, but I guess the idea was not to change this (not changing the time for releasing coq-serapi!), but just pushing another tag beforehand (that would not a "release of coq-serapi expected to work with both Coq 8.14+rc1 and Coq 8.14.0", but just a "tagged prerelease of coq-serapi" intended to be further tested with Coq 8.14+rc1.

Actually, as I mentioned in my previous comment, a mere SHA1 pin would be *almost* fine… though opam's git-pinning also requires a version string (in order to figure out the dependencies constraints w.r.t. other packages etc.), hence the idea to just push a tag for the commit at stake (which would then comes with a version string out of the box :)

**Comment by Zimmi48** [comment]
> Umm, maybe I am confused, but I see nothing in https://github.com/coq/ceps/blob/master/text/052-platform-release-cycle.md that requires stuff in the CI to do a tag or a release.

OK, the language is a bit unclear, and this is probably because of the inheritance from the former process where pins where done on the Coq side with commits. But still, see point 4:

>Starting with the pins made by RM on Coq's CI all packages part of the platform (or its core) are pinned (in accordance with upstreams, which are notified about the ongoing process).

And:

>Coq platform pins are communicated to upstream devs, which may tag/release.

I agree that the use of "may" isn't very prescriptive.

>On the contrary, the beta process works over pure pins from CI isn't it the case?

This CEP removed the beta process. The whole point of switching over to a release candidate is to indicate that: things should not break anymore, it is safe to tag now.

>Indeed, but I guess the idea was not to change this (not changing the time for you to release coq-serapi!), but just pushing another tag beforehand (that would not be a "release of coq-serapi expected to work with both Coq 8.14+rc1 and Coq 8.14.0", but just a "tagged prerelease of coq-serapi" intended to be further tested with Coq 8.14+rc1.

Well, the expectation is really that the tag for 8.14+rc1 should also work for 8.14.0 except in extenuating circumstances.

> > I personally don't get why coq-serapi is released in the main opam repo contrary to virtually all other Coq packages, but never mind that.
> 
> What would we gain from that?

Consistency, for one thing. But more importantly, it would make management of the package by interested parties easier. But now that SerAPI is there, it would be difficult to backtrack (as it would give the false idea that it has stopped releasing new versions).

**Comment by Zimmi48** [comment]
Thanks for all the feedback on the platform procedures. I do agree that they are not sufficiently clearly defined at the moment. Thus, I'll cc @MSoegtropIMC and @gares so they can join the discussion.

About the opam repository issue, it's much less important IMHO, but interested parties mostly being Coq users, they have more control on the Coq opam repository.

**Comment by ejgallego** [comment]
> About the opam repository issue, it's much less important IMHO, but interested parties mostly being Coq users, they have more control on the Coq opam repository.

Note that this problem is not specific to coq-serapi, in the sense that Coq also has package files in both Coq's OPAM and OCaml OPAM; so IMHO, there should be no impediment for coq-serapi to follow the same approach.

**Comment by ejgallego** [comment]
Ok, the https://github.com/ejgallego/coq-serapi/releases/tag/8.14%2Brc1%2B0.14.0 tag should work for Coq 8.14+rc1, where do I submit it?

Note that SerAPI 0.14 proper will contain quite a few foward ports of 0.13 features, so this is far from a final release.

File: sertop/sertop_init.ml [hunk 1]
```diff
@@ -46,11 +49,45 @@ type coq_opts = {
...
+
+  ()
+
 let coq_init opts out_fmt =
 
   if opts.debug then begin
     Printexc.record_backtrace true;
-    Flags.debug := true;
+    (* XXX Use CDebug *)
+    (* Flags.debug := true; *)
```
Comment (gares) [inline sertop/sertop_init.ml:90]:
```suggestion
    CDebug.set_debug_all true;
```

**Pull request closed without merging**